compiler/rustc_trait_selection/src/solve/mod.rs

```diff
@@ -383,6 +403,110 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
+                    };
+
+                    if let Some(substs_eq) = opt_substs_eq {
+                        // If we have `alias_eq(<T as Trait>::Assoc, <T as Trait>::Assoc)` we will end up here
+                        if lhs_to_rhs.is_err() && rhs_to_lhs.is_err() {
+                            return substs_eq;
+                        }
+
+                        // If we end up with `alias_eq(<?0 as Trait>::Assoc, <T as Trait>::Assoc)` we cant
+                        // unify `?0 = T` during coherence as there may exist some other concrete type for `?0`
+                        // that would normalize to `<T as Trait>::Assoc`.
+                        //
+                        // We don't want to return `substs_eq` in the case of it being `NoSolution` as this
+                        // goal may still be solvable by one of the `normalizes_to` goals.
+                        if (lhs_to_rhs.is_err() || rhs_to_lhs.is_err())
+                            && !self.infcx.intercrate
+                            && substs_eq.is_ok()
+                        {
+                            return substs_eq;
+                        }
+
+                        // If we relating the substs of the two aliases succeeds and introduces no constraints that
+                        // means we have something like `alias_eq(<?0 as Trait>::Assoc, <?0 as Trait>::Assoc)` and
+                        // should return a response of `Ok(Yes)` with no constraints regardless of what the `normalizes_to`
+                        // goals evaluated to.
+                        if let Ok(response) = &substs_eq {
+                            if response.value.certainty == Certainty::Yes
+                                && !response.has_inference_or_external_constraints()
+                            {
+                                return substs_eq;
+                            }
+                        }
+                    }
+
+                    match (lhs_to_rhs, rhs_to_lhs) {
+                        (Err(NoSolution), Err(NoSolution)) => Err(NoSolution),
+                        (Ok(response), Err(NoSolution)) | (Err(NoSolution), Ok(response)) => {
+                            Ok(response)
+                        }
+                        (Ok(response_lhs), Ok(response_rhs)) => {
+                            if response_lhs == response_rhs {
+                                Ok(response_lhs)
+                            } else {
+                                // FIXME(deferred_projection_equality): try taking the intersection of constraints
+                                // from each of the responses?
+                                self.make_canonical_response(
+                                    response_lhs
+                                        .value
+                                        .certainty
+                                        .unify_and(response_rhs.value.certainty),
+                                )
```

Comment:
I'm not sure it's correct to return an unconstrained response here? Should we just return ambiguous?